### PR TITLE
Added GameSystem and fixed PickupSystem

### PR DESCRIPTION
**Pull request by Bram-Boris**
# Description

This PR adds the GameSystem which will handle switching of levels and when the game is over.
I also implemented Random for choosing the level.

GCC needed an update to version 8 because 7 does not support `#include <filesystem>`

## Screenshot(s)
Very slow gif.
Very slow gif

# How can this be tested?

Use the random branch for the engine.
You can change the MAX LEVELS variable to limit the amount.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Included 1 or more screenshots
- [x] Code is const correct

# TODO
- [x] Players die when they leave the map.

### CODE REVIEW

File: lib/controllers/game_controller.cpp [hunk 1]
```diff
@@ -175,20 +180,42 @@ int GameController::getScreenHeight() const {
     return SCREEN_HEIGHT;
 }
 
-void GameController::startGame() {
+void GameController::startGame(int level_amount) {
```
Comment (MarkMeer) [inline lib/controllers/game_controller.cpp]: Level_amount is not used at this time, will it be used in the future or has it become obsolete during development?

File: lib/menu/main_menu.cpp [hunk 2]
```diff
@@ -39,7 +39,7 @@ MainMenu::MainMenu(int screen_width, int screen_height, GameController* game_con
     start_game_button.text.y_scale = 100;
     const auto game_controller_pass = game_controller;
     start_game_button.on_click = [game_controller_pass]() -> void {
-        game_controller_pass->startGame();
+        game_controller_pass->startGame(10);
```
Comment (MarkMeer) [inline lib/menu/main_menu.cpp]: Not something which should be done in this pull request, but would we like to (or should we) make this configurable while in the lobby or will this stay 10?

File: lib/scenes/scene_manager.cpp [hunk 3]
```diff
@@ -15,6 +15,16 @@
...
 void SceneManager::loadLevel(Level& level) {
+    // Create the players
+    current_scene_entities.push_back(entity_factory->createGorilla(-300, -300, 1));
+    current_scene_entities.push_back(entity_factory->createPanda1(-300, -300, 2));
+    current_scene_entities.push_back(entity_factory->createPanda2(-300, -300, 3));
+    current_scene_entities.push_back(entity_factory->createPanda3(-300, -300, 4));
+    current_scene_entities.push_back(entity_factory->createWeapon(1000, 200, true));
+    current_scene_entities.push_back(entity_factory->createWeapon(1100, 200, false));
+    current_scene_entities.push_back(entity_factory->createWeapon(600, 200, true));
+    current_scene_entities.push_back(entity_factory->createWeapon(500, 200, false));
```
Comment (MarkMeer) [inline lib/scenes/scene_manager.cpp]: The values are placeholders before they get spawned in the level which the user is about to play? 

Also, with how it is  done now, the players will be in the scene and get deleted. Isn't this the opposite to the idea we had about separating the players from the scene and 'recycling' them?

File: include/systems/game_system.hpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+#include "brickengine/components/player_component.hpp"
+#include "components/health_component.hpp"
+
+class GameSystem : public System {
+public:
+    GameSystem(std::shared_ptr<EntityManager> em, GameController& gc);
+    void update(double deltatime);
+private:
+    static constexpr int MAX_LEVELS = 4;
```
Comment (LuukSantegoeds) [inline include/systems/game_system.hpp]: ~This is 10 by default~
~What is this int for?~
After reading the rest of the changes I think this should be 10. This is the default amount of levels, right?

File: lib/systems/game_system.cpp [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    std::set<int> dead_players;
+    for (auto player : players) {
+        auto health = entity_manager->getComponent<HealthComponent>(player.first);
+        if (health->health <= 0 && !dead_players.count(player.first)) {
+            dead_players.insert(player.first);
+        }
+    }
+
+    int count =  players.size() - dead_players.size();
+    if(count == 1) {
```
Comment (MarkMeer) [inline lib/systems/game_system.cpp]: Is there a case in which the last two can kill each other and where <= 1 should be needed to verify the end of the game?

**Review by MarkMeer — DISMISSED** [review]

**Review by LuukSantegoeds — CHANGES_REQUESTED** [review]

File: lib/systems/game_system.cpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        // load next scene.
+        if (completed_level_amount <= MAX_LEVELS) {
+            game_controller.loadNextLevel();
+            ++completed_level_amount;
+        } else {
+            game_controller.loadMainMenu();
+            completed_level_amount = 0;
```
Comment (MarkMeer) [inline lib/systems/game_system.cpp]: When this gets a reset to 0, the next round will consist of 5 rounds instead of the 4 that is set in the game system.

**Review by MarkMeer — COMMENTED** [review]
I accidentally clicked approve, I would also like to request a few changes.

File: include/systems/game_system.hpp (hunk 4 above)
Comment (Bram-Boris) [reply include/systems/game_system.hpp]: Whoops, did not mean to commit this.

File: lib/menu/main_menu.cpp (hunk 2 above)
Comment (Bram-Boris) [reply lib/menu/main_menu.cpp]: This should be configurable within the lobby however, the main menu has no ability to reach the Game system and it not able to change the max levels which why I made it a default of 10. It is not a requirement to have different values for levels.

Comment (Bram-Boris) [reply lib/menu/main_menu.cpp]: So I think we skip the lobby level functionality and just make it 10.

File: lib/scenes/scene_manager.cpp (hunk 3 above)
Comment (Bram-Boris) [reply lib/scenes/scene_manager.cpp]: Then a player needs a reset function to reset their position, health, sizes etc to default.

File: lib/systems/game_system.cpp (hunk 5 above)
Comment (Bram-Boris) [reply lib/systems/game_system.cpp]: Yes that could happen. Would be very rare.

File: lib/menu/main_menu.cpp (hunk 2 above)
Comment (MarkMeer) [reply lib/menu/main_menu.cpp]: Completely agree, we should not over engineer these things when not necessary.

**Review by LuukSantegoeds — DISMISSED** [review]

**Review by MarkMeer — APPROVED** [review]
👍

**Review by MarkMeer — APPROVED** [review]

**Review by Peter-JanGootzen — APPROVED** [review]
welp

**Pull request merged**